pkg/updater/github.go

```diff
@@ -357,57 +357,74 @@ func (g *Github) ReplaceRunning(ctx context.Context, newBinary string) error {
+// getUncompressedReaderForArchive returns a io.ReadCloser that is the uncompressed contents of the archive
+func (g *Github) getUncompressedReaderForArchive(f *os.File) (io.Reader, func() error, error) {
+	if strings.HasSuffix(f.Name(), ".gz") {
+		gzr, err := gzip.NewReader(f)
+		return gzr, gzr.Close, err
+	} else if strings.HasSuffix(f.Name(), ".xz") {
+		// Use a buffered reader to speed up extraction:
+		// See: https://github.com/ulikunitz/xz/issues/23
+		xzr, err := xz.NewReader(bufio.NewReader(f))
+		return xzr, func() error { return nil }, err
+	}
+
+	return nil, nil, fmt.Errorf("unsupported archive type: %s", filepath.Ext(f.Name()))
+}
+
+// getFileFromArchive extracts a given file from the provided tar archive, returning its location
+func (g *Github) getFileFromArchive(ctx context.Context, f *os.File, storageDir, filenameInArchive string) (string, error) {
 	ctx = trace.StartCall(ctx, "github.getFileFromArchive")
 	defer trace.EndCall(ctx)
 
-	gzr, err := gzip.NewReader(f)
+	ucr, closer, err := g.getUncompressedReaderForArchive(f)
 	if err != nil {
-		return "", errors.Wrap(err, "failed to gzip decode stream")
+		return "", errors.Wrap(err, "failed to get uncompressed reader")
 	}
+	defer closer() //nolint:errcheck // Why: Best effort
 
-	tarReader := tar.NewReader(gzr)
-
-	srcFile, srcSize, err := findTarFile(ctx, tarReader, filename)
+	tarReader := tar.NewReader(ucr)
+	srcFile, srcSize, err := findFileInTarFile(ctx, tarReader, filenameInArchive)
 	if err != nil {
 		return "", err
 	}
 
-	file := filepath.Join(storageDir, filename)
-	targetFile, err := os.Create(file)
+	file := filepath.Join(storageDir, filenameInArchive)
+	destFile, err := os.Create(file)
 	if err != nil {
 		return "", err
 	}
-	defer targetFile.Close()
+	defer destFile.Close()
 
-	if g.Silent {
-		//nolint:gosec
-		_, err = io.Copy(targetFile, srcFile)
-	} else {
+	writer := io.Writer(destFile)
+	if !g.Silent {
+		// If we're not silent, use a multiwriter for the progress bar
+		// and output
 		bar := progressbar.DefaultBytes(
 			srcSize,
 			// extra space here to match the downloading update length
 			"extracting update ",
 		)
-		_, err = io.Copy(io.MultiWriter(targetFile, bar), srcFile) //nolint:gosec // wtaf?
+		writer = io.MultiWriter(destFile, bar)
 	}
-	if err != nil && !errors.Is(err, io.EOF) {
+
+	if _, err := io.Copy(writer, srcFile); err != nil && !errors.Is(err, io.EOF) {
 		return "", err
 	}
 
 	return file, nil
 }
 
-// finds specified file in provided tar archive and returns io.Reader and its size for extraction progress
-func findTarFile(ctx context.Context, archive *tar.Reader, filename string) (io.Reader, int64, error) {
+// findFileInTarFile finds a specified file in provided tar archive and returns io.Reader
+// and its size for extraction progress
+func findFileInTarFile(ctx context.Context, archive *tar.Reader, filename string) (io.Reader, int64, error) {
```

Comment:
How does this handle subdirs?

Reply:
It's like file.Watch, it doesn't need to handle them it'll just be walked over recursively